lib/facter/util/resolvers/http.rb

```diff
@@ -57,7 +57,19 @@ def make_request(url, headers, timeouts, request_type)
           end
 
           def http_obj(parsed_url, timeouts)
-            http = Net::HTTP.new(parsed_url.host)
+            # If Facter is querying a cloud metadata service and a proxy has
+            # been configured by the http_proxy environment variable, set
+            # Net::HTTP#new's p_addr argument to nil to allow the metadata
+            # service to properly resolve.
+            # Net::HTTP#new takes positional arguments, so we have to
+            # explicitly set all arguments up to and including p_addr.
+            http =
+              if ENV['http_proxy'] && parsed_url.to_s.match?(Facter::Resolvers::Ec2::EC2_METADATA_ROOT_URL || Facter::Resolvers::Ec2::EC2_USERDATA_ROOT_URL || Facter::Resolvers::Gce::METADATA_URL)
```

Comment:
I think it'd be cleaner to expose `http_host` and `http_port` as keyword arguments to the `get_request` and `put_request` methods and have the various cloud resolvers pass in `nil` values for those (always).